File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    ans = []
    policy = {}

    def to_day(year, month, day):
        return year * 12 * 28 + month * 28 + day

    today = to_day(*map(int, today.split('.')))
    for term in terms:
        name, mm = term.split()
        policy[name] = int(mm)

    for i, e in enumerate(privacies):
        validate_until, p = e.split()
        validate_until = to_day(*map(int, validate_until.split('.'))) + policy[p] * 28
        if validate_until <= today:
            ans.append(i + 1)
    return ans

File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;

class Solution {

	public int[] solution(String today, String[] terms, String[] privacies) {
		List<Integer> ans = new ArrayList<>();
		int todayInDays = convertDateToDays(today, "\\.");

		System.out.println("today : " + todayInDays);

		Map<String, Integer> validUntil = new HashMap<>();

		for (String term: terms) {
			String[] s = term.split(" ");
			validUntil.put(s[0], Integer.parseInt(s[1]));
		}

		for (int i = 0; i < privacies.length; i++) {
			String[] s1 = privacies[i].split(" ");
			int curDay = convertDateToDays(s1[0], "\\.");
			if (curDay + validUntil.get(s1[1]) * 28 <= todayInDays) ans.add(i + 1);

			System.out.println("curDay : " + curDay + ", validDays" + validUntil.get(s1[1]) + " total : " + curDay + validUntil.get(s1[1]));
		}


		return ans.stream().mapToInt(i -> i).toArray();
	}

	private int convertDateToDays(String date, String regex){
		String[] days = date.split(regex);
		return Integer.parseInt(days[0]) * 12 * 28 + Integer.parseInt(days[1]) * 28 + Integer.parseInt(days[2]);
	}
}